src/Bandits/BanditEvaluator.php

```diff
@@ -0,0 +1,251 @@
+    ): float {
+        $score = $coefficients->intercept;
+
+        $score += self::scoreNumericAttributes(
+            $coefficients->subjectNumericCoefficients,
+            $subjectAttributes->numericAttributes
+        );
+        $score += self::scoreCategoricalAttributes(
+            $coefficients->subjectCategoricalCoefficients,
+            $subjectAttributes->categoricalAttributes
+        );
+        $score += self::scoreNumericAttributes(
+            $coefficients->actionNumericCoefficients,
+            $actionAttributes->numericAttributes
+        );
+        $score += self::scoreCategoricalAttributes(
+            $coefficients->actionCategoricalCoefficients,
+            $actionAttributes->categoricalAttributes
+        );
+
+        return $score;
+    }
+
+    /**
+     * @param array<string, float> $actionScores
+     * @param float $gamma
+     * @param float $probabilityFloor
+     * @return array<string, float>
+     */
+    public static function weighActions(array $actionScores, float $gamma, float $probabilityFloor): array
+    {
+        $numberOfActions = count($actionScores);
+        $bestActionKey = array_keys($actionScores, max($actionScores))[0];
+
+
+        $minProbability = $probabilityFloor / $numberOfActions;
+
+        $weights = array_filter(
+            $actionScores,
+            function ($key) use ($bestActionKey) {
+                return $key !== $bestActionKey;
+            },
+            ARRAY_FILTER_USE_KEY
+        );
+
+        $bestScore = $actionScores[$bestActionKey];
+        $weights = array_map(
+            function ($score) use ($minProbability, $bestScore, $gamma, $numberOfActions) {
+                return max($minProbability, 1.0 / ($numberOfActions + $gamma * ($bestScore - $score)));
+            },
+            $weights
+        );
+
+        $remainingWeight = max(0.0, 1.0 - array_sum($weights));
+        $weights[$bestActionKey] = $remainingWeight;
+
+        return $weights;
+    }
+
+    /**
+     * @param string $flagKey
+     * @param string $subjectKey
+     * @param array<string, float> $actionWeights
+     * @return string
+     * @throws BanditEvaluationException
+     */
+    private function selectAction(string $flagKey, string $subjectKey, array $actionWeights): string
+    {
+        // Use a list of key-value pairs to make sorting easier.
+        $weightPairs = array_map(
+            fn($key) => new ActionValue($key, $actionWeights[$key]),
+            array_keys($actionWeights)
+        );
+
+        // usort sorts in place.
+        usort(
+            $weightPairs,
+            function (ActionValue $a, ActionValue $b) use ($subjectKey, $flagKey) {
+                $aValue = Sharder::getShard("$flagKey-$subjectKey-{$a->action}", $this->totalShards);
+                $bValue = Sharder::getShard("$flagKey-$subjectKey-{$b->action}", $this->totalShards);
+
+                return $aValue === $bValue ? 0 : (($aValue < $bValue) ? -1 : 1);
```

Comment:
We need to tie break scores with action names alphabetically

Also, for this ternary, I think you can just return $aValue - $bValue as `usort()` uses only the _sign_ of the return value ([docs](https://www.php.net/manual/en/function.usort.php))

Reply:
what a fruitful morning of reading docs! I am sheepish to admit I'd never seen the [_spaceship operator_](https://wiki.php.net/rfc/combined-comparison-operator) before.